rfc/0017-policy-server-reuse-wasm-optimized-modules.md

```diff
@@ -0,0 +1,292 @@
+>
+> [...] If this `Hash` matches between two `Engine`s then binaries from one are guaranteed to deserialize in the other.
+
+When optimizing an Wasm module, we could also store the hash of the `Engine` that
+produced it.
+
+Later, we can use this hash to quickly find a pre-optimized Wasm module that we
+could reuse.
+
+## Where to store the pre-optimized modules
+
+Right now each `PolicyServer` CR is implemented by using a `Deployment`. The
+Deployment usually has a replica size of 2, but this detail can be changed by
+the user.
+All the `Pod`s that are part of the `Deployment` are storing their data inside
+of a temporary directory. The contents of this directory are lost whenever
+the Pod is restarted (like during a Deployment rollout).
+
+We could change the `Deployment` template to ensure all its Pods have access
+to a Persistent Volume. We could then use this PV to store both vanilla and
+the optimized Wasm modules.
+
+In this scenario, we have to find a way to ensure only one Pod has
+exclusive write access to the directory where the vanilla and the optimized Wasm
+files are going to be written.
+
+This could be done using a file lock. However, file locks do not work in
+a consistent way across different network file systems. This could lead to
+some race conditions happening with certain combinations of PV drivers.
+We fear that using a file lock would cause a higher maintenance.
+
+Another possibility is to use a Kubernetes primitive called
+[`Lease`](https://kubernetes.io/docs/concepts/architecture/leases/) to ensure
+that only one of the Pods (the leader) has write access to the PV.
+
+## Using Kubernetes `Lease`
+
+The `Lease` resource is a primitive leveraged by Kubernetes controllers to build
+a lock system that is then used for the leader election process.
+
+The hard job of building a lock system is done on the client side. For example,
+the Kubernetes Go client does that inside of the
+[`k8s.io/client-go/tools/leaderelection/resourcelock`](https://pkg.go.dev/k8s.io/client-go@v0.27.3/tools/leaderelection/resourcelock)
+package.
+In the rust ecosystem this functionality is provided by the [kubert](https://crates.io/crates/kubert)
+crate. This crate is also used by one of [linkerd controllers](https://github.com/linkerd/linkerd2/tree/main/policy-controller).
+
+We can rely on [`kubert::lease::LeaseManager`](https://docs.rs/kubert/0.16.1/kubert/lease/struct.LeaseManager.html)
+to ensure exclusive write access to this shared Persistent Volume.
+
+## New deployment model
+
+Policy Server would continue to be deployed using a Deployment object.
+A new Persistent Volume would then be defined. This is where all the `.wasm`
+files and all their optimized versions would be saved.
+
+The deployment would then define a new init container. The Policy Server
+configuration would be injected into the init container as a read-only file.
+The container would also have write access to the Persistent Volume described
+above.
```

Comment:
In a multi-node scenario, this only works if the storage class supports the `ReadWriteMany`. Otherwise, the volume is not shared between nodes.
Being `ReadWriteMany` is only supported by specific filesystems, the user should be aware of this.
Another approach could be having one volume per pod which will let us eliminate the lease. We could suggest to the user to isolate slow startup policies to a different policy server.